### PR TITLE
Implement OCSP response signing

### PR DESCRIPTION
**Pull request by bifurcation**
CAs need to sign OCSP responses in addition to certificates.  This PR is a draft of the CLI for signing OCSP responses, with a stubbed-out OCSP module.

Does this look like it's headed in roughly the right direction?  I'm currently not doing the local/remote/universal split that is done in the signer module.

### CODE REVIEW

**Comment by grittygrease** [comment]
There are some conflicts to fix, but in general, this is going in the right direction. We can add the remote/universal signers and the ocspsign API as a later PR.

File: cli/cli.go [hunk 1]
```diff
@@ -215,3 +226,14 @@ func PrintCert(key, csrBytes, cert []byte) {
 	}
 	fmt.Printf("%s\n", jsonOut)
 }
+
+// PrintOcspResponse outputs an OCSP response to stdout
+func PrintOcspResponse(resp []byte) {
```
Comment (grittygrease) [inline cli/cli.go]: Go style dictates this should be PrintOCSPResponse.

**Comment by grittygrease** [comment]
You might want to wait for https://github.com/cloudflare/cfssl/pull/95 to be resolved before revisiting this.

File: cli/cli.go [hunk 2]
```diff
@@ -94,6 +101,11 @@ func registerFlags(c *Config, f *flag.FlagSet) {
 	f.StringVar(&c.IP, "ip", "", "remote server ip")
 	f.StringVar(&c.Remote, "remote", "", "remote CFSSL server")
 	f.StringVar(&c.Label, "label", "", "key label to use in remote CFSSL server")
+	f.StringVar(&c.ResponderFile, "responder", "", "Certificate for OCSP responder")
+	f.StringVar(&c.Status, "status", "good", "Status of the certificate: good, revoked, unknown")
+	f.IntVar(&c.Reason, "reason", 0, "Reason code for revocation")
+	f.StringVar(&c.RevokedAt, "revoked-at", "now", "Date of revocation (YYYY-MM-DD)")
+	f.IntVar(&c.Interval, "interval", 3600*24*4, "Interval between OCSP updates, in seconds (default: 4 days)")
```
Comment (kisom) [inline cli/cli.go]: The default value here should probably be a constant, like that for [`OneYear`](https://github.com/cloudflare/cfssl/blob/55a735b51c216b71e87dda61b74e7770b476cfdb/helpers/helpers.go#L20) in helpers.

File: cli/ocspsign/ocspsign.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+		Certificate: cert,
+		Status:      c.Status,
+	}
+
+	if c.Status == "revoked" {
+		req.Reason = c.Reason
+
+		req.RevokedAt = time.Now()
+		if c.RevokedAt != "now" {
+			req.RevokedAt, err = time.Parse("2001-01-01", c.RevokedAt)
```
Comment (kisom) [inline cli/ocspsign/ocspsign.go]: Should this be 2001-01-02?

File: ocsp/ocsp.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+	issuer    *x509.Certificate
+	responder *x509.Certificate
+	key       crypto.Signer
+	interval  time.Duration
+}
+
+// NewSignerFromFile reads the issuer cert, the responder cert and the responder key
+// from PEM files, and takes an interval in seconds
+func NewSignerFromFile(issuerFile, responderFile, keyFile string, interval int) (Signer, error) {
+	log.Debug("Loading CA cert: ", issuerFile)
```
Comment (kisom) [inline ocsp/ocsp.go]: Maybe "Loading OCSP cert" instead?

Comment (kisom) [inline ocsp/ocsp.go]: I guess the better option would be "Loading responder cert" to keep it consistent with the debug logs below.

File: ocsp/ocsp.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+		interval:  time.Duration(interval) * time.Second,
+	}, nil
+}
+
+func (s StandardSigner) Sign(req SignRequest) ([]byte, error) {
+	if req.Certificate == nil {
+		return nil, errors.New("req.Certificate == nil") // XXX
+	}
+
+	// Verify that req.Certificate is issued under s.issuer
```
Comment (kisom) [inline ocsp/ocsp.go]: Maybe we should add a new cfssl/errors category for OCSP.

**Comment by grittygrease** [comment]
@bifurcation Can you address the comments so we can merge?

File: cli/cli.go (hunk 2 above)
Comment (bifurcation) [inline cli/cli.go]: Changed to "4 * helpers.OneDay", which seems like it strikes the right balance between readability and constant specificity.

File: cli/ocspsign/ocspsign.go (hunk 3 above)
Comment (bifurcation) [inline cli/ocspsign/ocspsign.go]: Actually, should be "2006-01-02" in order to encode the reference date.

File: ocsp/ocsp.go (hunk 4 above)
Comment (bifurcation) [inline ocsp/ocsp.go]: This one is loading the *issuer* cert, the one for whom we're signing OCSP responses.  The responder cert is below.  Changed to "issuer cert".

File: ocsp/ocsp.go (hunk 5 above)
Comment (bifurcation) [inline ocsp/ocsp.go]: Added, and migrated this file to use it.

**Comment by kisom** [comment]
LGTM.

**Pull request merged**